### DISCUSSION TITLE
Would "one-click apps" be a good idea?

### DISCUSSION THREAD

**Discussion by benallfree** [pockethost/pockethost#218] [Q&A]
[This discussion](https://github.com/benallfree/awesome-pocketbase/pull/20#issuecomment-1582052953) inspired an idea for supporting the ability to create PocketHost projects that have pre-configured assets (migrations, data, static files).

I'm curious to hear from all of you, does this seem useful? As PocketBase-powered applications proliferate, it might be really cool to be able to deploy a backend for the app in one click on PocketHost.

Example: I am a developer and have made a realtime chat app named PocketChat. I wrote an Android client, an iOS client, and a desktop client. PocketChat allows users to point to their own PocketBase-powered backend so they have a completely private chat platform. Everything is open source. As part of this project, I submit (and maintain) a PR to PocketHost for a one-click app named PocketChat. It contains default migrations and static assets required to support my app. In my readme.md, I tell users that they can either set up their own PocketBase with migrations and assets, or they can go to PocketHost and provision a PocketChat backend instantly.

I could also envision preconfigured/boilerplate setups being helpful as starting points for various types of applications. For example, you could have a `PocketBlog` app that contained `posts`, `comments`, etc, all complete with the appropriate security rules. Or `PocketShop` that has `products`, `orders`, `reviews`, etc, along with stock assets like sample product images and such.

What does anyone think? 

@ganigeorgiev do you have any specific thoughts on whether this might be helpful to users? I'd probably approach it as a separate repo like `awesome-pocketbase` so it wouldn't necessarily be tied to PocketHost, so the community could contribute and maintain it for other purposes as well. Or maybe it could even just be a new section on `awesome-pocketbase`?

edit: There actually might be one PocketHost-specific characteristic of the app templates. PocketHost allows Deno workers for cloud function support. If people included these, it would be a PocketHost-specific backend.

edit 2: another approach would be to allow instances to be publicly discoverable and cloneable. So I could maintain the base version of an app and users could clone their own instances from mine. Hmm, I think I like the monorepo idea better though. PH could just source its app catalog from a separate monorepo.

**Comment by ganigeorgiev** [comment]
> @ganigeorgiev do you have any specific thoughts on whether this might be helpful to users? 

I guess the app templates could be useful as starter/skeleton type of thing, but it really depends on what the template will include.

If the template is supposed to be only collections configurations (schema fields, rules, indexes, etc.) the easiest way would be to import an already exported collections JSON into the Admin UI (_or automate the process using the [Import collections API](https://pocketbase.io/docs/api-collections/)_).

For anything more complex I imagine that it may require some special handling in PocketHost when creating the instance to pull the app template files from somewhere (migrations, `pb_public` files, , custom workers code, etc.; _note that with the next release there will be also basic support for js app hooks_).

**Comment by davidbarton** [comment]
Perhaps @ganigeorgiev will correct me. But If I make an assumption that full app state can be recreated via migrations. Then this feature is already possible with migrations.

Yet I would dare to make the argument that one click apps would improve the newcomer experience and draw more signups and people testing out PocketHost/PocketBase.

In order to make one click apps usable for production I think some form of https://github.com/pocketbase/pocketbase/issues/718 and documentation around it is needed. Image you one click deploy a service to manage collection of data for your users. Yet they still need to go through your full auth flow instead of using existing native auth.

**Comment by davidbarton** [comment]
> Perhaps @ganigeorgiev will correct me. But If I make an assumption that full app state can be recreated via migrations. Then this feature is already possible with migrations.

This is also one way to implement those one-click apps. And as a fan of gitops a preferred one for me. Dealing with `pb_data` is "dirty" and it asks the developer to interact with internal implementation details. Crafting a set of migrations to build target state is much cleaner way to do it. It can be easily reviewed and audited as well.

PocketBase already offers ways to generate some of these migrations for you. Perhaps if more such "export state to migration" would be offered you could in theory build the entire application in UI (the UI is so great btw) and export it using migrations to git repo that would seve as one click app for others.

**Comment by benallfree** [comment]
https://pocodex.dev exists now.
